### DISCUSSION TITLE
Is it possible to create a "grid" layout programmatically with add panel?

### DISCUSSION THREAD

**Discussion by mikedidthis** [mathuo/dockview#420] [Q&A] [answered]
<img width="1500" alt="Screenshot 2024-01-03 at 15 46 43" src="https://github.com/mathuo/dockview/assets/910412/79058289-486d-4840-b4ef-82ff48b6ce28">

Firstly thanks for the awesome library! 

Is it possible to create the above layout using the api only (I can do it via drag and drop)? I may have missed something in the docs.

**Comment by mathuo** [comment, marked as answer]
Glad you like it! 

There are some docs on the `addPanel` method you may find useful here

https://dockview.dev/docs/next/components/dockview#add-panel

To briefly explain, you can provide a `position` object which takes a couple of arguments which help position the panel. You can position the new panel relative to an existing panel by providing `referencePanel` or an existing group by providing `referenceGroup`. You can also provide a `direction` which will be relative to the `referencePanel` or `referenceGroup` if provided and if not provided the `direction` is absolute.

[Here](https://codesandbox.io/p/sandbox/elastic-nash-ctjq29?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqxyzjgq0006356j8evcmugp%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqxyzjgq0002356js9xus7ll%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqxyzjgq0003356jvv497ods%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqxyzjgq0005356j9gpr2tnd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqxyzjgq0002356js9xus7ll%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqxyzjgq0001356jnleia87t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clqxyzjgq0002356js9xus7ll%2522%252C%2522activeTabId%2522%253A%2522clqxyzjgq0001356jnleia87t%2522%257D%252C%2522clqxyzjgq0005356j9gpr2tnd%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqxyzjgq0004356jb21d1qkz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clqxyzjgq0005356j9gpr2tnd%2522%252C%2522activeTabId%2522%253A%2522clqxyzjgq0004356jb21d1qkz%2522%257D%252C%2522clqxyzjgq0003356jvv497ods%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clqxyzjgq0003356jvv497ods%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) is a quick sandbox showing the layout you mention created using `addPanel` and a variety of positioning options, you could achieve the same layout with many alternative positioning configurations but this is one of them.

Let me know if that helps and if you have any other questions.

**Reply by mikedidthis** [reply]
Thanks for the quick response. I think the piece I was missing is tracking the reference panel and a single add panel.

I will try and create an MVP and report back.
